text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,980 @@
+region graph in various common scenarios, and in the end falls back to
+enforcing conditions that may be stricter than necessary, but which
+certainly suffice. We have not yet encountered an example in practice
+where the current implementation rules do not suffice.
+
+### The WF relation
+
+This section describes the "well-formed" relation. In
+[previous RFCs][RFC 192], this was combined with the outlives
+relation. We separate it here for reasons that shall become clear when
+we discuss WF conditions on impls.
+
+The WF relation is really pretty simple: it just says that a type is
+"self-consistent". Typically, this would include validating scoping
+(i.e., that you don't refer to a type parameter `X` if you didn't
+declare one), but we'll take those basic conditions for granted.
+
+    WfScalar:
+      --------------------------------------------------
+      R ⊢ scalar WF
+
+    WfParameter:
+      --------------------------------------------------
+      R ⊢ X WF                  // where X is a type parameter
+
+    WfTuple:
+      ∀i. R ⊢ Ti WF
+      --------------------------------------------------
+      R ⊢ (T0..Tn) WF
+
+    WfNominalType:
+      ∀i. R ⊢ Pi Wf             // parameters must be WF,
+      C = WhereClauses(Id)      // and the conditions declared on Id must hold...
+      R ⊢ [P0..Pn] C            // ...after substituting parameters, of course
+      --------------------------------------------------
+      R ⊢ Id<P0..Pn> WF
+
+    WfReference:
+      R ⊢ T WF                  // T must be WF
+      R ⊢ T: 'x                 // T must outlive 'x
+      --------------------------------------------------
+      R ⊢ &'x T WF
+
+    WfSlice:
+      R ⊢ T WF
+      R ⊢ T: Sized
+      --------------------------------------------------
+      [T] WF
+
+    WfProjection:
+      ∀i. R ⊢ Pi WF             // all components well-formed
+      R ⊢ <P0: Trait<P1..Pn>>   // the projection itself is valid
+      --------------------------------------------------
+      R ⊢ <P0 as Trait<P1..Pn>>::Id WF
+
+#### WF checking and higher-ranked types
+
+There are two places in Rust where types can introduce lifetime names
+into scope: fns and trait objects. These have somewhat different rules
+than the rest, simply because they modify the set `R` of bound
+lifetime names. Let's start with the rule for fn types:
+
+    WfFn:
+      ∀i. R, r.. ⊢ Ti WF
+      --------------------------------------------------
+      R ⊢ for<r..> fn(T1..Tn) -> T0 WF
+
+Basically, this rule says that a `fn` type is *always* WF, regardless
+of what types it references.  This certainly accepts a type like
+`for<'a> fn(x: &'a T)`.  However, it also accepts some types that it
+probably shouldn't. Consider for example if we had a type like
+`NoHash` that is not hashable; in that case, it'd be nice if
+`fn(HashMap<NoHash, u32>)` were not considered well-formed. But these
+rules would accept it, because `HashMap<NoHash,u32>` appears inside a
+fn signature.
```

Comment:
I'm probably blind here, but doesn't the rule check that all argument types are WF? Following the `NoHash` example, for `HashMap<NoHash, u32>` to be WF, the `where` clause is checked according to the definitions above.

Also, assuming the explanation is true and I somehow misunderstood the inference rule, why is it a good idea to accept such types (and presumably catch such errors later, when the function is actually called - where does this manifest)?